core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java

```diff
@@ -153,33 +151,66 @@ private static void add(Map<String, Settings> executorSettings, ExecutorSettings
         executorSettings.put(name, settings);
     }
 
-    private static class ExecutorSettingsBuilder {
-        Map<String, String> settings = new HashMap<>();
+    private static abstract class ExecutorSettingsBuilder<T extends ExecutorSettingsBuilder<T>> {
 
-        public ExecutorSettingsBuilder(String name) {
+        private final Map<String, String> settings;
+
+        protected ExecutorSettingsBuilder(String name, ThreadPoolType threadPoolType) {
+            if (THREAD_POOL_TYPES.get(name) != threadPoolType) {
+                throw new IllegalArgumentException("thread pool [" + name + "] must be of type [" + threadPoolType + "]");
```

Comment:
I'm +1 on this but it's new, right? Can we clearly document it in the PR description and title?